### ISSUE TITLE
Better XML Stanza and error support

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11241]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#1686|
|Type|enhancement|
|Created|2006-04-30 19:04:33Z|

The `twisted.words.protocols.jabber.client` module features an IQ class that models an `<iq/>` XML Stanza. However, this implementation is limited to the `jabber:client` namespace. Also, it would be very nice if worked with deferreds.

Support for `<message/>` and `<presence/>` XML Stanzas.

Also, support for errors (both stream level and stanza level) should be added:

  * Parsing of error elements
  * The ability to generate an exception object from an incoming
    error stanza
  * Have exception objects generate an error response stanza from
    the incoming stanza that caused the exception

<details><summary>Searchable metadata</summary>

```
trac-id__1686 1686
type__enhancement enhancement
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_4 Words-0.4
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1146423873000000 1146423873000000
changetime__1147701440000000 1147701440000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm removed owner|
|-|-|

At this point, I see no immediate need for `<message/>` or `<presence/>` support.

The rest has been implemented in [branches/xmpp-stanzas-1686.](branches/xmpp-stanzas-1686.) Please review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @ralphm|
|-|-|

From Pyflakes:
```
twisted/words/test/test_jabberxmlstream.py:3: 'defer' imported but unused
twisted/words/protocols/jabber/xmlstream.py:21: 'STREAM_CONNECTED_EVENT' imported but unused
twisted/words/protocols/jabber/xmlstream.py:22: 'STREAM_START_EVENT' imported but unused
twisted/words/protocols/jabber/xmlstream.py:23: 'STREAM_END_EVENT' imported but unused
```

 * Line 18 of test_jabberxmlstream violates 80 cols
 * Line 146 of test_jabbererror violates 80 cols
 * It's not clear at a glance what _behaviour_ ExceptionFromStanzaTest is supposed to be testing

Otherwise, looks good.  Do a full test run and merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set status to `closed`|
|-|-|

(In [16830]) Add stanza and stream error support, new IQ stanza helper.

Author: ralphm
Reviewer: jml
Fixes [#11241](https://github.com/twisted/twisted/issues/11241)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
